Libraries/rocSOLVER/syev/main.cpp

```diff
@@ -0,0 +1,167 @@
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "cmdparser.hpp"
+#include "example_utils.hpp"
+#include "rocblas_utils.hpp"
+
+#include <rocblas/rocblas.h>
+#include <rocsolver/rocsolver.h>
+
+#include <hip/hip_runtime.h>
+
+#include <iomanip>
+#include <iostream>
+#include <numeric>
+#include <random>
+#include <vector>
+
+int main(const int argc, char* argv[])
+{
+    // 1. Parse user input
+    cli::Parser parser(argc, argv);
+    parser.set_optional<rocblas_int>("n", "n", 3, "Size of n x n input matrix A");
+    parser.run_and_exit_if_error();
+
+    // Get the n x n matrix size
+    const rocblas_int n = parser.get<rocblas_int>("n");
+    if(n <= 0)
+    {
+        std::cout << "Value of 'n' should be greater than 0" << std::endl;
+        return error_exit_code;
+    }
+    const rocblas_int lda = n;
+
+    // 2. Data vectors
+    std::vector<rocblas_double> A(n * lda); // Input matrix
+    std::vector<rocblas_double> V(n * lda); // Resulting eigenvectors
+    std::vector<rocblas_double> W(n); // Resulting eigenvalues
+
+    // 3. Generate a random symmetric matrix
+    std::default_random_engine                     generator;
+    std::uniform_real_distribution<rocblas_double> distribution(0., 2.);
+    auto random_number = [&]() { return distribution(generator); };
+
+    for(int i = 0; i < n; i++)
+    {
+        A[(n + 1) * i] = random_number();
+        for(int j = 0; j < i; j++)
+        {
+            A[i * n + j] = A[j * n + i] = random_number();
+        }
+    }
+
+    // 4. Set rocSOLVER parameters
+    const rocblas_evect evect = rocblas_evect::rocblas_evect_original;
+    const rocblas_fill  uplo  = rocblas_fill::rocblas_fill_lower;
+
+    // 5. Reserve and copy data to device
+    rocblas_double* d_A    = nullptr;
+    rocblas_double* d_W    = nullptr;
+    rocblas_int*    d_info = nullptr;
+
+    HIP_CHECK(hipMalloc(&d_info, sizeof(rocblas_int)));
+    HIP_CHECK(hipMalloc(&d_A, sizeof(rocblas_double) * A.size()));
+    HIP_CHECK(hipMalloc(&d_W, sizeof(rocblas_double) * W.size()));
+    HIP_CHECK(hipMemcpy(d_A, A.data(), sizeof(rocblas_double) * A.size(), hipMemcpyHostToDevice));
+
+    // 6. Initialize rocBLAS.
+    rocblas_handle handle;
+    ROCBLAS_CHECK(rocblas_create_handle(&handle));
+
+    // 7. Get and reserve the working space on device.
+    rocblas_int     d_work_size = n * 3 - 1;
```

Comment:
Actually, there is an open discussion about this https://github.com/ROCmSoftwarePlatform/rocSOLVER/issues/543. Supposedly the workspace should have enough space to store a tridiagonal matrix (used internally in the algorithm) so that is at least `n + n - 1 + n - 1` = `3 * n - 2` elements. However, as mentioned in the discussion, when [testing syev](https://github.com/ROCmSoftwarePlatform/rocSOLVER/blob/1a4498cf91c68f147efdf70868f28c4b489b8fc0/clients/include/testing_syev_heev.hpp#L302) they use `3 * n - 1` for the workspace size, so I think it's ok to use this value for now.

Reply:
I should probably have myself reassigned to that issue, since it seems like it hasn't had any movement for a while.

The tridiagonal matrix is internally stored in arrays D and E (called dW and d_work here). D contains the `n` diagonal elements (which converge to the eigenvalues), and E contains the `n-1` off-diagonal elements (which converge to zero; it's a symmetric tridiagonal matrix, so only one off-diagonal is needed). E (aka d_work) of size `n` should be sufficient for the function to work; if not, then that's a bug and I'll need to take a look at it. Using `3*n-1` as the size isn't wrong, per se, but it is using more device memory than what is needed.

Reply:
Yes, it would be great if we could close that issue. Okay so based on this new information we could as well use `n - 1` for the `d_work` size, and that is supposedly the minimum size needed. I feel like it would be best for the examples to use this value instead of `n`, because this way noob users of this library can be made aware of the fact that at least `n - 1` is needed.

PD: it would also be great if could have this information added to the documentation.

Reply:
Changes implemented in the last commit :)